File: user/m_files/plot_g4_neutron_xs.m
function [ ] = plot_g4_neutron_xs( Z )
%function [ ] = plot_g4_neutron_xs( Z, color = 'blue', prev_handles = [], keV_tf = true )

% [pes, pis, pc] = plot_g4_neutron_xs( Z, keV_tf = true ) -- plot GEANT4 neutron-interaction cross-sections
%   for given element
% -- [pes, pis, pc] are the graphics handles for the elastic-scattering plot,
%      the inelastic-scattering plot, and the capture plot, respectively
% -- Z is the atomic number of the desired nucleus
% -- keV_tf: if true, changes energy units to keV; otherwise, defaults to MeV
% -- optional isotope "I" argument (equal to isotope nuclear mass in amu)
%      may be added in the future
% ~ Mark J. Duvall ~ mjduvall@hawaii.edu ~ 6/21 ~ %


%Copyright (C) 2021 Mark J. Duvall
%
%    This program is free software: you can redistribute it and/or modify
%    it under the terms of the GNU General Public License as published by
%    the Free Software Foundation, either version 3 of the License, or
%    (at your option) any later version.
%
%    This program is distributed in the hope that it will be useful,
%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%    GNU General Public License for more details.
%
%    You should have received a copy of the GNU General Public License
%    along with this program.  If not, see <http://www.gnu.org/licenses/>.

% filesystem init
[status, datadir] = system("echo -n $G4NEUTRONXSDATA");
if isempty(datadir)
  disp "Error: Environment variable $G4NEUTRONXSDATA not found. Were environment scripts sourced?"
  return
endif
[status2, datadir_basename] = system("basename $G4NEUTRONXSDATA");

% prepare axis items
en_lim = [1e-9 1e5];
en_label = 'Neutron Energy (MeV)';
en_therm = (1/40)*1e-6;
en_reactor_low = 1e-3; % K_n = 1 keV
en_reactor_high = 1e-2; % K_n = 10 keV
xs_lim = [1e-30 1e-18];
% switch to keV
en_lim = en_lim * 1e3;
en_label = 'Neutron Energy (keV)';
en_therm = en_therm*1e3;
en_reactor_low = en_reactor_low*1e3;
en_reactor_high = en_reactor_high*1e3;

% plot prep
f = figure;
ax = axes('xlim', en_lim, 'ylim', xs_lim, 'xscale', 'log', 'yscale', 'log');
set(gcf, 'name', 'GEANT4 Neutron Cross-Section Data');
hold on
l_es = line( [en_lim(1) en_lim(1)], [xs_lim(1) xs_lim(1)], 'color', 'black' );
l_is = line( [en_lim(1) en_lim(1)], [xs_lim(1) xs_lim(1)], 'linestyle', ':', 'color', 'black' );
l_cap = line( [en_lim(1) en_lim(1)], [xs_lim(1) xs_lim(1)], 'linestyle', '--', 'color', 'black' );
set( [l_es l_is l_cap], 'visible', 'off')
l_therm = line([en_therm en_therm], [xs_lim(1)*.1 xs_lim(2)*10], 'color', 'black', 'linewidth', 1, 'linestyle', ':');
p = patch( [en_reactor_low en_reactor_high en_reactor_high en_reactor_low], [xs_lim(2)*10 xs_lim(2)*10 xs_lim(1)*.1 xs_lim(1)*.1], 'facecolor', [.5 .5 .5], 'facealpha', 0.2, 'edgecolor', 'none');
legend_items = [l_es l_is l_cap l_therm p];
legend_names = {'Elastic Scattering', 'Inelastic Scattering', 'Capture', 'Thermal Energy', 'Reactor Region'};

% labels
Tstr = sprintf("Neutron Cross Sections from %s", datadir_basename);
T = title(Tstr);
xlabel(en_label)
ylabel 'Cross Section (cm^{2})'
set(ax, 'fontsize', 16)
colorlist = 'brmgc';
icolor = 1;

% MAIN
for z = Z
  % read data
  capfile = sprintf("%s/cap%d", datadir, z);
  elastfile = sprintf("%s/elast%d", datadir, z);
  inelastfile = sprintf("%s/inelast%d", datadir, z);
  cap = dlmread(capfile, '', 2, 0);
  el = dlmread(elastfile, '', 2, 0);
  inel = dlmread(inelastfile, '', 2, 0);
  % switch to keV
  el(:,1) = el(:,1)*1e3;
  inel(:,1) = inel(:,1)*1e3;
  cap(:,1) = cap(:,1)*1e3;
  % remove non-plottable data
  cap(isnan(cap(:,2)),:) = [];
  el(isnan(el(:,2)),:) = [];
  inel(isnan(inel(:,2)),:) = [];
  cap(cap(:,2)<=0,:) = [];
  el(el(:,2)<=0,:) = [];
  inel(inel(:,2)<=0,:) = [];
  % make plots
  pes = plot(el(:,1), el(:,2), 'linewidth', 2);
  pes_label = sprintf('Z = %d', z);
  set(gca, 'ylim', xs_lim)
  pis = plot(inel(:,1), inel(:,2), 'linewidth', 2, 'linestyle', ':');
  pc = plot(cap(:,1), cap(:,2), 'linewidth', 2, 'linestyle', '--');
  set( [pes pis pc], 'color', sprintf('%s', colorlist(icolor)))
  icolor += 1;
  legend_items(end+1) = pes;
  legend_names(end+1) = pes_label;
end
%set(ax, 'xscale', 'log')
%set(ax, 'yscale', 'log')

% legend
l = legend( legend_items, legend_names, 'location', 'northeastoutside' );

% adjust axes
%set(gca, 'xlim', en_lim)
%set(gca, 'ylim', xs_lim)
% workaround
disp "To fix the axes, run the following command: set(gca, 'ylim', [1e-30 1e-18])"

% all pau!   )
%endfunction

File: user/root_macros/util/plotG4NeutronXS.cxx
// plotG4NeutronXS -- plot neutron cross-sections from $G4NEUTRONXS data
// ~ Mark J. Duvall ~ mjduvall@hawaii.edu ~ 7/2021 ~ //


//void plotG4NeutronXS( Int_t Z, const enum EColor = kBlue ) {
void plotG4NeutronXS( Int_t Z ) {


// filesystem setup
TString g4xs_absolute = gSystem->Getenv("G4NEUTRONXSDATA");
TString g4xs_file = g4xs_absolute( g4xs_absolute.Last('/')+1, g4xs_absolute.Length()-1 );
if (g4xs_absolute == "") {
  gSystem->Error("Getenv", "Error: Environment variable $G4NEUTRONXSDATA not found. Were environment scripts sourced?\n");
  return;
}
TString elastFile, inelastFile, capFile;
elastFile = TString::Format("%s/elast%d", g4xs_absolute.Data(), Z);
inelastFile = TString::Format("%s/inelast%d", g4xs_absolute.Data(), Z);
capFile = TString::Format("%s/cap%d", g4xs_absolute.Data(), Z);

//debug
//printf("%s %s %s %s %s\n", g4xs_absolute.Data(), g4xs_file.Data(), elastFile.Data(), inelastFile.Data(), capFile.Data());
//TString shellcmd = TString::Format("ls -l %s", elastFile.Data());
//gSystem->Exec(shellcmd.Data());

// read data into TTrees
TTree* T_e = new TTree( "T_e", "Elastic Scattering Data" );
TTree* T_i = new TTree( "T_i", "Inelastic Scattering Data" );
TTree* T_c = new TTree( "T_c", "Capture Data" );
T_e->ReadFile( elastFile.Data(), "energy/D:xs/D");
T_i->ReadFile( inelastFile.Data(), "energy/D:xs/D");
T_c->ReadFile( capFile.Data(), "energy/D:xs/D");

// prepare axes
TString enLabel = "Energy (keV)";
TString xsLabel = "Cross-Section (cm^{2})";
Double_t en_low, en_high, en_therm, en_reactor_low, en_reactor_high, xs_low, xs_high;
en_low = 1.e-6; en_high = 1.e2;
en_therm = (1/40)*1e-3;
en_reactor_low = 1.e-6; en_reactor_high = 1.e-5;
xs_low = 1.e-30; xs_high = 1.e-18;


// MAIN
TCanvas* c = new TCanvas("c", g4xs_file.Data());
//TH2D* he = new TH2D;
//TH2D* hi = new TH2D;
//TH2D* hc = new TH2D;
T_e->Draw("xs:energy>>he", "xs<1");
//he->GetXaxis()->SetLimits(en_low, en_high);
//he->GetYaxis()->SetLimits(xs_low, xs_high);
//he->SetMarkerStyle(2);
//he->Draw();
//T_i->Draw("xs:energy>>hi", "xs<1", "csame");
//T_c->Draw("xs:energy>>hc", "xs<1", "csame");
c->SetLogx(true);
c->SetLogy(true);




// all pau!   )
}
